### PR TITLE
Use MCP h/w polarity function for inverting inputs

### PR DESCRIPTION
**Pull request by moinmoin-sh**
Don't know if the merge will work, tells me "Can't automatically merge.  Don’t worry, you can still create the pull request. "
Don't understand Git .....

This is what I wanted to say : 

Here is a new proposal:

A few days ago there was a discussion on using the build-in HW invert function of the MCP.
I took a closer look and came to this proposal

The invert (polarity) function of the MCP can easily be activated. 
There is no need to do this at compile time (NO hard coding!). The invert bit can be set and reset on the fly (see new function MCPsetInvert()).

I see the following advantages:
utilize the HW that we have.
receiving the conf/invt message sets the invert bit in the MCP and it becomes active immediately
no bookkeeping of the invert variable required - like fire and forget.
saves resources (flash and ram)
the button code sees active input always as low
the port animation on the OLED shows active inputs always with the little square
there is no user observable difference to the sw solution 

What do you think

### CODE REVIEW

File: Firmware/USMArduinoFirmware/USM_Input.cpp [hunk 1]
```diff
@@ -18,7 +18,6 @@ USM_Input::USM_Input()
   {
     // Default all inputs to buttons (non-inverted)
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/USM_Input.cpp:19]: Fix comment

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino [hunk 2]
```diff
@@ -443,12 +443,12 @@ void mqttCallback(char * topic, byte * payload, int length)
     
     if (length == 0 || strncmp((char*)payload, "0", length) == 0)
     {
-      usmInput[mcp].setInvert(input, 0);
+      MCPsetInvert(mcp, input, 0);
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/USMArduinoFirmware.ino:446]: Could we rename this to `setMcpInvert()`? And maybe rename the `val` parameter to `invert` so it is clear what that is related to?

File: Firmware/USMArduinoFirmware/USMArduinoFirmware.ino [hunk 3]
```diff
@@ -829,3 +830,38 @@ byte readRegister(byte r)
...
+    uint8_t reg = (pin < 8) ? MCP23017_IPOLA : MCP23017_IPOLB;  
+    
+    uint8_t tmp = MCPreadReg(adr, reg);
+    bitWrite(tmp, pin % 8, val);
+    MCPwriteReg(adr, reg, tmp);
+  }
+}
+
+// helper : read 1 byte from MCP register
+uint8_t MCPreadReg(int adr, uint8_t reg)
```
Comment (sumnerboy12) [inline Firmware/USMArduinoFirmware/USMArduinoFirmware.ino:851]: We already have `readRegister()` at the bottom of the sketch, for reading the MAC ROM register. I think we can modify that to pass the register address and use it for both the MAC and MCP register reads.

Then rename `MCPwriteReg` to `writeRegister()` to match the existing `readRegister()` naming format.

I.e. turn these into generic I2C register read/write helpers.

**Review by sumnerboy12 — CHANGES_REQUESTED** [review]
I agree with your comments - not sure it is any simpler in terms of coding, but it is simpler in terms of state we need to store/maintain and as you say, performs exactly the same function with less memory requirements.

**Comment by sumnerboy12** [comment]
BTW - I have fixed the merge conflicts, which were caused by my pushing a few changes yesterday after you created your branch. In future if you want to avoid that sort of thing you need to rebase your PR branch to the branch you are wanting to merge to.

**Comment by moinmoin-sh** [comment]
Thanks for your comments and suggestions. Wasn't sure if you would like it.
I'll take care for it a little later. Have plumbers coming in today and tomorrow.

**Comment by sumnerboy12** [comment]
I do kind of like the fact the inversion logic was encapsulated in the
USM_Input library, alongside the type config, but I can see the value in
using the h/w features on offer.

Plus the memory savings can't hurt!

**Comment by moinmoin-sh** [comment]
I think I made the changes that you suggested.

Please check if the reading of the MAC-address rom still works - I don't have that HW

I made an addional change in the mqtt-callback. I think it's required to do a range check of the index in the topic because if that number is out of range a false mcp index would be calculated which leads to a write to a wrong address. That could cause weired behavior depending where the write goes.

**Comment by sumnerboy12** [comment]
I just pushed a couple of minor changes - everything worked fine on my test system but I just wanted to tidy up a few little bits. Thanks for the PR!

**Pull request merged**